sphinx4-core/src/main/java/edu/cmu/sphinx/api/FeatureExtractor.java

```diff
@@ -0,0 +1,146 @@
+import edu.cmu.sphinx.frontend.FloatData;
+import edu.cmu.sphinx.frontend.FrontEnd;
+import edu.cmu.sphinx.frontend.util.StreamDataSource;
+import edu.cmu.sphinx.util.props.ConfigurationManager;
+
+/**
+ * Extracts features from input stream
+ * 
+ * @author Vladisav Jelisavcic
+ * 
+ */
+class FeatureExtractor {
+    private FrontEnd frontEnd;
+    private StreamDataSource audioSource;
+    private int featureLength = -1;
+
+    /** The logger for this class */
+    private static final Logger logger = Logger
+            .getLogger("edu.cmu.sphinx.api.FeatureExtractor");
+
+    /**
+     * Constructs a FeatureExtractor. 
+     * 
+     * @param cm
+     *            the configuration manager
+     * @param frontEndName
+     *            the name for the frontEnd to be used
+     * @param inputStream
+     * 			  data stream 
+     *            
+     * @throws IOException if error occurred
+     */
+    public FeatureExtractor(ConfigurationManager cm, String frontEndName, InputStream inputStream)
+            throws IOException {
+    
+        if(cm.lookup(frontEndName) == null) {
+        	throw new RuntimeException("No such frontend: " + frontEndName);
+        }
+        
+        frontEnd = (FrontEnd) cm.lookup(frontEndName);
+        audioSource = (StreamDataSource) cm.lookup("streamDataSource");
+        audioSource.setInputStream(inputStream);
+    }
+	
+    /**
+     * Extracts all features from the supplied InputStream. 
+     * 
+     * @return float[][] when called first time, null otherwise 
+     * 
+     * @throws DataProcessinException if error occurred
+     */
+	public float[][] getAllFeatures() throws DataProcessingException {
+		List<float[]> featureList = new ArrayList<float[]>();
+	
+	    Data feature = frontEnd.getData();
+	    if(feature == null)
+	    	return null;
+	    
+        while (!(feature instanceof DataEndSignal)) {
+            if (feature instanceof DoubleData) {
+                double[] featureData = ((DoubleData) feature).getValues();
+                if (featureLength < 0) {
+                    featureLength = featureData.length;
+                    logger.info("Feature length: " + featureLength);
+                }
+                float[] convertedData = new float[featureData.length];
+                for (int i = 0; i < featureData.length; i++) {
+                    convertedData[i] = (float) featureData[i];
+                }
+                featureList.add(convertedData);
+            } else if (feature instanceof FloatData) {
+                float[] featureData = ((FloatData) feature).getValues();
+                if (featureLength < 0) {
+                    featureLength = featureData.length;
+                    logger.info("Feature length: " + featureLength);
+                }
+                featureList.add(featureData);
+            }
+            feature = frontEnd.getData();
+        }
+
+	     float[][] allFeatures = new float[featureList.size()][];
```

Comment:
It is ok to return feature list as list, not necessary an array. It's better to make sure code is properly indented too.